agent/taskresource/credentialspec/credentialspec_windows.go

```diff
@@ -0,0 +1,600 @@
+	// needed mostly for testing.
+	ssmClientCreator ssmfactory.SSMClientCreator
+
+	// s3ClientCreator is a factory interface that creates new S3 clients. This is
+	// needed mostly for testing.
+	s3ClientCreator s3factory.S3ClientCreator
+
+	// credentialSpecResourceLocation is the location for all the tasks' credentialspec artifacts
+	credentialSpecResourceLocation string
+
+	// required for processing credentialspecs, key is input credentialspec
+	// Example key := credentialspec:file://credentialspec.json
+	requiredCredentialSpecs map[string][]*apicontainer.Container
+
+	// map to transform credentialspec values, key is a input credentialspec
+	// Examples:
+	// * key := credentialspec:file://credentialspec.json, value := credentialspec=file://credentialspec.json
+	// * key := credentialspec:s3ARN, value := credentialspec=file://CredentialSpecResourceLocation/s3_taskARN_fileName.json
+	// * key := credentialspec:ssmARN, value := credentialspec=file://CredentialSpecResourceLocation/ssm_taskARN_param.json
+	credSpecMap map[string]string
+
+	// lock is used for fields that are accessed and updated concurrently
+	lock sync.RWMutex
+}
+
+// NewCredentialSpecResource creates a new CredentialSpecResource object
+func NewCredentialSpecResource(taskARN, region string,
+	credentialSpecs map[string][]*apicontainer.Container,
+	executionCredentialsID string,
+	credentialsManager credentials.Manager,
+	ssmClientCreator ssmfactory.SSMClientCreator,
+	s3ClientCreator s3factory.S3ClientCreator) (*CredentialSpecResource, error) {
+
+	s := &CredentialSpecResource{
+		taskARN:                 taskARN,
+		region:                  region,
+		requiredCredentialSpecs: credentialSpecs,
+		credentialsManager:      credentialsManager,
+		executionCredentialsID:  executionCredentialsID,
+		ssmClientCreator:        ssmClientCreator,
+		s3ClientCreator:         s3ClientCreator,
+		credSpecMap:             make(map[string]string),
+	}
+
+	err := s.setCredentialSpecResourceLocation()
+	if err != nil {
+		return nil, err
+	}
+
+	s.initStatusToTransition()
+	return s, nil
+}
+
+func (cs *CredentialSpecResource) initStatusToTransition() {
+	resourceStatusToTransitionFunction := map[resourcestatus.ResourceStatus]func() error{
+		resourcestatus.ResourceStatus(CredentialSpecCreated): cs.Create,
+	}
+	cs.resourceStatusToTransitionFunction = resourceStatusToTransitionFunction
+}
+
+func (cs *CredentialSpecResource) Initialize(resourceFields *taskresource.ResourceFields,
+	taskKnownStatus status.TaskStatus,
+	taskDesiredStatus status.TaskStatus) {
+	cs.initStatusToTransition()
+	cs.credentialsManager = resourceFields.CredentialsManager
+	cs.ssmClientCreator = resourceFields.SSMClientCreator
+	cs.s3ClientCreator = resourceFields.S3ClientCreator
+
+	// if task hasn't turn to 'created' status, and it's desire status is 'running'
+	// the resource status needs to be reset to 'NONE' status so the cs value
+	// will be retrieved again
+	if taskKnownStatus < status.TaskCreated &&
```

Comment:
If the resource has already moved to created, that means the s3 files would have been downloaded. in this case, they will be downloaded again. how are we planning to handle this? ideally, the files would have to be cleaned up in some way.

Reply:
As per the design, it is alright to download the same file multiple times.

Reply:
My question here is do we need to download it again? So for secret, we need to call ASM/SSM API to retrieve secrets again is bcuz they are all sensitive data and we don't want to preserve the data in our data file. However, based on your case, it is just a file name. So you should be able to directly save it to data file and restore it when agent restarts. It is not rly necessary to download it again.

Reply:
it is possible that someone removes the files in between.. we could either check if the required files are present and then try to download it again. 
What's concerning is multiple downloads.

Reply:
>  it is alright to download the same file multiple times.

why do you think so?

Reply:
discussed offline: to avoid inconsistencies, we are not downloading resources that are already.

Reply:
Changed as discussed.